tests/storage/test_room_search.py

```diff
@@ -0,0 +1,82 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import os
+
+import synapse.rest.admin
+from synapse.rest.client import login, room
+
+from tests.unittest import HomeserverTestCase
+
+
+class NullByteInsertionTest(HomeserverTestCase):
+    servlets = [
+        synapse.rest.admin.register_servlets_for_client_rest_resource,
+        login.register_servlets,
+        room.register_servlets,
+    ]
+
+    def test_null_byte(self):
+        """
+        Postgres/SQLite don't like null bytes going into the search tables. Internally
+        we replace those with a space.
+
+        Ensure this doesn't break anything.q
+
+        """
+
+        # register a user and create a room, create some messages
+        self.register_user("alice", "password")
+        access_token = self.login("alice", "password")
+        room_id = self.helper.create_room_as("alice", True, "1", access_token)
+        body1 = "hi\u0000bob"
+        body2 = "another message"
+        body3 = "hi alice"
+
+        # send messages and ensure they don't cause an internal server
+        # error
+        resp1 = self.helper.send(room_id, body1, "1", access_token)
+        resp2 = self.helper.send(room_id, body2, "2", access_token)
+        resp3 = self.helper.send(room_id, body3, "3", access_token)
+        self.assertTrue("event_id" in resp1)
+        self.assertTrue("event_id" in resp2)
+        self.assertTrue("event_id" in resp3)
+
+        # check that search still works with the message where the null byte was replaced
+        store = self.hs.get_datastore()
+        res1 = self.get_success(
+            store.search_msgs([room_id], "hi bob", ["content.body"])
+        )
+        self.assertEquals(res1.get("count"), 1)
+
+        # check that search still works with another unrelated message
+        res2 = self.get_success(
+            store.search_msgs([room_id], "another", ["content.body"])
+        )
+        self.assertEquals(res2.get("count"), 1)
+
+        # check that search still works when given a search term that overlaps
+        # with the message that we replaced the null byte in and an unrelated one
+        res3 = self.get_success(store.search_msgs([room_id], "hi", ["content.body"]))
+        self.assertEquals(res3.get("count"), 2)
+        res4 = self.get_success(
+            store.search_msgs([room_id], "hi alice", ["content.body"])
+        )
+
+        # check content of highlights if we are using postgres
+        if "SYNAPSE_POSTGRES" in os.environ and os.environ["SYNAPSE_POSTGRES"] == "1":
```

Comment:
I would probably prefer to check that our store engine is Postgres directly, rather than rely on these env vars being present — *just in case* we ever change these around and don't remember to fix it here.

I think this will do it, but may be wrong:

```suggestion
        if isinstance(store.database_engine, PostgresEngine):
```